src/key-backup.ts

```diff
@@ -0,0 +1,188 @@
+			scope: 'https://www.googleapis.com/auth/devstorage.full_control',
+			aud: 'https://oauth2.googleapis.com/token',
+			exp: expires,
+			iat: now,
+		};
+
+		const jwt = await JWT.sign(payload, serviceAccount.private_key, { algorithm: 'RS256' });
+
+		const response = await fetch('https://oauth2.googleapis.com/token', {
+			redirect: 'manual',
+			method: 'POST',
+			headers: {
+				'Content-Type': 'application/x-www-form-urlencoded',
+			},
+			body: new URLSearchParams({
+				grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
+				assertion: jwt,
+			}).toString(),
+		});
+
+		if (!response.ok) {
+			const errorText = await response.text();
+			logger.error('Google OAuth Token Error:', errorText);
+			throw new Error(
+				`Failed to get Google Access Token: ${response.status} ${response.statusText} - ${errorText}`
+			);
+		}
+
+		const data: { access_token: string; expires_in: number } = await response.json();
+
+		// Cache the new token with a slight buffer before actual expiry
+		cachedAccessToken = {
+			token: data.access_token,
+			expiry: Date.now() + (data.expires_in - 60) * 1000, // 60 seconds buffer
+		};
+
+		return data.access_token;
+	} catch (e) {
+		logger.error('Error during Google Access Token generation:', e);
+		throw new Error('Authentication failed.');
+	}
+};
+
+type Key = {
+	id: string;
+	privateKey: Uint8Array;
+	metadata: Record<string, string>;
+};
+
+type Required<T> = { [K in keyof T]: NonNullable<T[K]> };
+
+type BackupEnv = Required<Pick<Bindings, 'BACKUPS_SERVICE_ACCOUNT_KEY' | 'BACKUPS_BUCKET_NAME'>>;
+
+const backupKeys = async (logger: WshimLogger, env: BackupEnv, keys: Key[]): Promise<void> => {
+	const gcsUploadUrl = `https://storage.googleapis.com/upload/storage/v1/b/${env.BACKUPS_BUCKET_NAME}/o?uploadType=multipart`;
+	const boundary = 'metadata-boundary';
+
+	const objectNamePrefix = (() => {
+		const now = new Date();
+		return `${now.getUTCFullYear()}-${now.getUTCMonth()}-${now.getUTCDate()}`;
+	})();
+
+	const accessToken = await getGoogleAccessToken(logger, env.BACKUPS_SERVICE_ACCOUNT_KEY);
+	for (const { id, privateKey, metadata } of keys) {
+		const objectName = `${objectNamePrefix}-id:${id}`;
+		const objMetadata = {
+			name: objectName,
+			metadata: Object.fromEntries(
+				Object.entries(metadata).map(([k, v]) => [`x-goog-meta-${k}`, v])
+			),
+		};
+
+		const post = {
+			redirect: 'manual',
+			method: 'POST',
+			headers: {
+				'Authorization': `Bearer ${accessToken}`,
+				'Content-Type': `multipart/related; boundary=${boundary}`,
+			},
+			body: [
+				`--${boundary}`,
+				'Content-Type: application/json; charset=UTF-8',
+				'',
+				JSON.stringify(objMetadata),
+				'',
+				`--${boundary}`,
+				'Content-Type: application/octet-stream',
+				'Content-Transfer-Encoding: base64',
+				'',
+				u8ToB64(privateKey),
+				`--${boundary}--`,
+				'',
+			].join('\r\n'),
+		};
+		try {
+			const uploadResponse = await fetch(gcsUploadUrl, post);
```

Comment:
I think we should try ~3-5 times with backoff + randomized jitter. Another option is to place the backup tasks in a CF queue. My concern here is what happens if GCS is down. Will we try to backup everything again later?